internal/contentmanager/cache/cache.go

```diff
@@ -0,0 +1,240 @@
+package cache
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type Watcher interface {
+	Watch(source.Source) error
+}
+
+// Cache is a storage mechanism for keeping track of
+// managed content sources
+type Cache interface {
+	io.Closer
+	// Watch establishes watches for all provided client.Objects.
+	// Subsequent calls to Watch will result in no longer necessary
+	// watches being stopped and removed and new watches being
+	// created
+	Watch(context.Context, Watcher, ...client.Object) error
+}
+
+// CloserSyncingSource is a wrapper of the controller-runtime
+// source.SyncingSource that includes methods for:
+//   - Closing the source, stopping it's interaction with the Kubernetes API server and reaction to events
+type CloserSyncingSource interface {
+	source.SyncingSource
+	io.Closer
+}
+
+type sourcerer interface {
+	Source(schema.GroupVersionKind, client.Object, func(context.Context)) (CloserSyncingSource, error)
+}
+
+type cache struct {
+	sources     map[schema.GroupVersionKind]CloserSyncingSource
+	sourcerer   sourcerer
+	owner       client.Object
+	syncTimeout time.Duration
+	mu          sync.Mutex
+}
+
+func NewCache(sourcerer sourcerer, owner client.Object, syncTimeout time.Duration) Cache {
+	return &cache{
+		sources:     make(map[schema.GroupVersionKind]CloserSyncingSource),
+		sourcerer:   sourcerer,
+		owner:       owner,
+		syncTimeout: syncTimeout,
+	}
+}
+
+var _ Cache = (*cache)(nil)
+
+func (c *cache) Watch(ctx context.Context, watcher Watcher, objs ...client.Object) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	gvkSet, err := gvksForObjects(objs...)
+	if err != nil {
+		return fmt.Errorf("getting set of GVKs for managed objects: %w", err)
+	}
+
+	if err := c.removeStaleSources(gvkSet); err != nil {
+		return fmt.Errorf("removing stale sources: %w", err)
+	}
+	return c.startNewSources(ctx, gvkSet, watcher)
+}
+
+func (c *cache) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	errs := []error{}
+	for _, source := range c.sources {
```

Comment:
Map iteration is randomly ordered. Should we figure out a way to sort the errors that pop out to avoid these propagating and flapping a Kubernetes object status?

Reply:
Errors sorted in https://github.com/operator-framework/operator-controller/pull/1119/commits/f6f277035a8331898fd304504b30660905797406